### PR TITLE
Rework JsonUtils' optional handling to let Converters see null

### PR DESCRIPTION
**Pull request by DHowett**
The JsonUtils changes in #8018 revealed that we need more robust,
configurable optional handling. We learned that there's a class of
values that was previously underrepresented in our API: _strings that
have an explicit empty value_.

The Settings model supports starting directory, icon, background image
et al values that are empty. That emptiness _overrides_ a value set in a
lower layer, so it is not sufficient to represent the empty value for
any one of those fields as an unset optional.

There are a couple other settings for which we've implemented a
hand-rolled option type (for roughly the same reason): foreground,
background, any color fields that override values from the color scheme
_or_ the lower layer profile.

These requirements are best fulfilled by better optional support in
JsonUtils. Where the library would originally detect known types of
optional and pre-filter them out during `GetValue` and `SetValue`, it
will now defer to another conversion trait.

This commit introduces a helper conversion trait and an "option oracle".
The conversion trait will use the option oracle to detect emptiness,
generate empty option values, and read values out of option types. In so
doing, the trait is insulated from the implementation details of any
specific option type.

Any special logic for handling JSON null and option types has been
stripped from GetValue. Due to this, there is an express change in
behavior for some converters:

* `GetValue<T>(jsonNull)` where `T` is **not** an option type[1] has
  been upgraded from a silent no-op to an exception.

Further, I took the opportunity to replace NullableSetting with
std::optional<std::optional<T>>, which accurately represents "setting
that the user might explicitly clear". I've added a test to
JsonUtilsTests to make sure it can serialize/deserialize double
optionals the way we expect it to.

Tests (Local, Unit for TerminalApp/SettingsModel):
Summary: Total=140, Passed=140, Failed=0, Blocked=0, Not Run=0, Skipped=0

[1]: Explicitly, if `T` is not an option type _and the converter does
not support null_.

### CODE REVIEW

File: src/cascadia/TerminalSettingsModel/JsonUtils.h [hunk 1]
```diff
@@ -177,13 +175,27 @@ namespace Microsoft::Terminal::Settings::Model::JsonUtils
...
+            }
             return winrt::hstring{ til::u8u16(Detail::GetStringView(json)) };
         }
 
         Json::Value ToJson(const winrt::hstring& val)
         {
+            if (val == winrt::hstring{})
+            {
+                return Json::Value::nullSingleton();
+            }
```
Comment (carlos-zamora) [inline src/cascadia/TerminalSettingsModel/JsonUtils.h:190]: _Interesting_. So we store `null` as `""` but write it out as `null`. I like that.

File: src/cascadia/ut_app/JsonUtilsTests.cpp [hunk 2]
```diff
@@ -101,6 +101,38 @@ JSON_FLAG_MAPPER(JsonTestFlags)
     };
 };
 
+struct hstring_like
+{
+    std::string value;
+};
```
Comment (carlos-zamora) [inline src/cascadia/ut_app/JsonUtilsTests.cpp:107]: What? Why?

**Review by carlos-zamora — APPROVED** [review]

File: src/cascadia/TerminalSettingsModel/JsonUtils.h (hunk 1 above)
Comment (DHowett) [reply src/cascadia/TerminalSettingsModel/JsonUtils.h:190]: I mean, this is extremely exactly how hstring works. We've talked about this, I think: _THERE IS NO EMPTY HSTRING_. There is an hstring that contains text, or an hstring that is null. There is no `""` hstring.

File: src/cascadia/ut_app/JsonUtilsTests.cpp (hunk 2 above)
Comment (DHowett) [reply src/cascadia/ut_app/JsonUtilsTests.cpp:107]: I mean, was there something in the test that made this unclear?

File: src/cascadia/TerminalSettingsModel/IInheritable.h [hunk 3]
```diff
@@ -72,11 +72,7 @@ namespace winrt::Microsoft::Terminal::Settings::Model::implementation
 
     // This is like std::optional, but we can use it in inheritance to determine whether the user explicitly cleared it
     template<typename T>
-    struct NullableSetting
-    {
-        std::optional<T> setting{ std::nullopt };
-        bool set{ false };
-    };
+    using NullableSetting = std::optional<std::optional<T>>;
```
Comment (carlos-zamora) [inline src/cascadia/TerminalSettingsModel/IInheritable.h:75]: @miniksa will have a laugh here

File: src/cascadia/TerminalSettingsModel/IInheritable.h [hunk 4]
```diff
@@ -148,51 +144,51 @@ public:                                                                     \
...
     {                                                                       \
         const auto val{ _get##name##Impl() };                               \
-        if (val.set)                                                        \
+        if (val)                                                            \
         {                                                                   \
-            if (val.setting)                                                \
+            if (*val)                                                       \
             {                                                               \
-                return *val.setting;                                        \
+                return **val;                                               \
```
Comment (carlos-zamora) [inline src/cascadia/TerminalSettingsModel/IInheritable.h:159]: hehe

**Comment by DHowett** [comment]
```
Summary of Non-passing Tests:
    TerminalAppLocalTests::TabTests::TryZoomPane [Failed]
    TerminalAppLocalTests::TabTests::MoveFocusFromZoomedPane [Failed]
    TerminalAppLocalTests::TabTests::CloseZoomedPane [Failed]

Summary: Total=143, Passed=140, Failed=3, Blocked=0, Not Run=0, Skipped=0
```

These test failures are not caused by this change.

**Review by carlos-zamora — APPROVED** [review]
I love this so much. Double approve haha

File: src/cascadia/TerminalSettingsModel/IInheritable.h (hunk 3 above)
Comment (miniksa) [reply src/cascadia/TerminalSettingsModel/IInheritable.h:75]: I thought we agreed this was gonna be `til::super_optional<T>` or `til::optional_no_really<T>`. :P

Comment (DHowett) [reply src/cascadia/TerminalSettingsModel/IInheritable.h:75]: If we have a need for it outside of TSM, absolutely ;)

**Comment by DHowett** [comment]
Since this changed the behavior of some more settings, it'll require @zadjii-msft to sign off. :smile:

File: src/cascadia/TerminalSettingsModel/IInheritable.h [hunk 5]
```diff
@@ -201,12 +197,12 @@ private:                                                                    \
         /*iterate through parents to find a value*/                         \
         for (auto parent : _parents)                                        \
```
Comment (DHowett) [inline src/cascadia/TerminalSettingsModel/IInheritable.h:198]:
```suggestion
        for (const auto& parent : _parents)                                 \
```

Comment (DHowett) [reply src/cascadia/TerminalSettingsModel/IInheritable.h:198]: idk how we missed this in code review

shouldn't we have had a helper for "walk parents and do a thing"? I thought that was part of the brief.

Comment (carlos-zamora) [reply src/cascadia/TerminalSettingsModel/IInheritable.h:198]: The problem I ran into by taking this out into its own helper method was that we're returning a value and calling a specific `_get##name##Impl()`. Figured it'd be less of a headache to throw this in a macro.

The `const &` thing is my bad tho. Sorry! I think I went through all of the other ones in the PR, but I just forgot to check the macro?

**Review by zadjii-msft — APPROVED** [review]
Okay, I feel good about this because there are good tests for this, not because I understand templates

**Pull request merged**